contractcourt/htlc_incoming_contest_resolver.go

```diff
@@ -78,6 +78,37 @@ func (h *htlcIncomingContestResolver) processFinalHtlcFail() error {
 	return nil
 }
 
+// Launch will call the inner resolver's launch method if the preimage can be
+// found, otherwise it's a no-op.
+func (h *htlcIncomingContestResolver) Launch() error {
+	// NOTE: we don't mark this resolver as launched as the inner resolver
+	// will set it when it's launched.
+	if h.isLaunched() {
+		h.log.Tracef("already launched")
+		return nil
+	}
+
+	h.log.Debugf("launching contest resolver...")
+
+	// Query the preimage and apply it if we already know it.
+	applied, err := h.findAndapplyPreimage()
+	if err != nil {
+		return err
+	}
+
+	// No preimage found, leave it to be handled by the resolver.
+	if !applied {
+		return nil
+	}
+
+	h.log.Debugf("found preimage for htlc=%x,  transforming into success "+
+		"resolver and launching it", h.htlc.RHash)
+
+	// Once we've applied the preimage, we'll launch the inner resolver to
+	// attempt to claim the HTLC.
+	return h.htlcSuccessResolver.Launch()
```

Comment:
> shouldn't we just morph directly into the success resolver?

So the morphing is done in the contest resolver's `Resolve`. Basically we have two goroutines - one for receiving blockbeat and `Launch` the resolver, the other for `Resolve` which morphs into the success resolver and `Launch` the new resolver. Since the resolver is protected via `launched` bool, launching it twice would be NOOP.